src/puppetlabs/puppetdb/query/catalogs.clj

```diff
@@ -1,106 +1,111 @@
-   :tags (s/maybe [String])
-   :exported (s/maybe s/Bool)
-   :file (s/maybe String)
-   :line (s/maybe s/Int)
-   :parameters (s/maybe String)
-   :source_type (s/maybe String)
-   :source_title (s/maybe String)
-   :target_type (s/maybe String)
-   :target_title (s/maybe String)
-   :relationship (s/maybe String)})
+   :resources (s/maybe PGobject)
+   :transaction_uuid (s/maybe s/Str)
+   :version (s/maybe s/Str)})
 
-(defn catalog-response-schema
-  "Returns the correct schema for the `version`, use :all for the full-catalog (superset)"
-  [api-version]
-  (case api-version
-    :v4 (assoc (cats/catalog-wireformat :v5) :hash String)
-    (cats/catalog-wireformat api-version)))
+;; MUNGE
 
-(defn create-catalog-pred
-  [rows]
-  (let [catalog-hash (:hash (first rows))]
-    (fn [row]
-      (= catalog-hash (:hash row)))))
+(pls/defn-validated rtj->resource :- catalogs/resource-query-schema
+  "Convert the row_to_json PG output to real data, and parse parameters
+  from its JSON storage."
+  [resource :- {s/Str s/Any}]
+  (-> resource
+      (set/rename-keys {"f1" :resource
+                        "f2" :type
+                        "f3" :title
+                        "f4" :tags
+                        "f5" :exported
+                        "f6" :file
+                        "f7" :line
+                        "f8" :parameters})
+      (update-in [:parameters] #(json/parse-strict-string % true))))
 
-(defn collapse-resources
-  [acc row]
-  (let [{:keys [tags type title line parameters exported file]} row
-        resource {:tags tags :type type :title title :line line
-                  :parameters (json/parse-strict-string parameters true)
-                  :exported exported :file file}]
-    (into acc
-          (-> resource
-              (kitchensink/dissoc-if-nil :line :file)
-              vector))))
+(pls/defn-validated resources->expansion :- catalogs/resources-expanded-query-schema
+  "Convert the resources data to a href/data style object."
+  [obj :- (s/maybe PGobject)
+   certname :- s/Str
+   base-url :- s/Str]
+  (let [data-obj {:href (str base-url "/catalogs/" certname "/resources")}]
+    (if obj
+      (assoc data-obj :data (map rtj->resource
+                                 (json/parse-string (.getValue obj))))
+      data-obj)))
 
-(defn collapse-edges
-  [acc row]
-  (let [{:keys [source_type target_type source_title target_title relationship]} row
-        edge {:source {:type source_type :title source_title}
-              :target {:type target_type :title target_title}
-              :relationship relationship}]
-    (into acc [edge])))
+(pls/defn-validated rtj->edge :- catalogs/edge-query-schema
+  "Convert the row_to_json PG output to real data."
+  [edge :- {s/Str s/Any}]
+  (-> edge
+      (set/rename-keys {"f1" :source_type
+                        "f2" :source_title
+                        "f3" :target_type
+                        "f4" :target_title
+                        "f5" :relationship})))
```

Comment:
this rename-keys call is one of the biggest time-consumers in catalog processing. I was able to cut the query time for catalogs by more than half by changing this code to use set/rename in edges->expansion instead of mapping this. The change would necessitate some schema alterations too since it outputs a set.